crates/ruma-identifiers/src/client_secret.rs

```diff
@@ -0,0 +1,172 @@
+//! Client secret identifier.
+
+use std::{convert::TryFrom, fmt, mem, rc::Rc, str::FromStr, sync::Arc};
+
+use ruma_identifiers_validation::client_secret::validate;
+
+/// A client secret.
+///
+/// Session IDs in Matrix are opaque character sequences of `[0-9a-zA-Z.=_-]. Their length must
+/// must not exceed 255 characters.
+#[repr(transparent)]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize), serde(transparent, crate = "serde"))]
+pub struct ClientSecret(str);
+
+/// An owned client secret.
+pub type ClientSecretBox = Box<ClientSecret>;
+
+impl ClientSecret {
+    #[allow(clippy::transmute_ptr_to_ptr)]
+    fn from_borrowed(s: &str) -> &Self {
+        unsafe { mem::transmute(s) }
+    }
+
+    fn from_owned(s: Box<str>) -> Box<Self> {
+        unsafe { mem::transmute(s) }
+    }
+
+    fn into_owned(self: Box<Self>) -> Box<str> {
+        unsafe { mem::transmute(self) }
+    }
+
+    /// Creates a string slice from this `ClientSecret`.
+    pub fn as_str(&self) -> &str {
+        &self.0
+    }
+
+    /// Creates a byte slice from this `ClientSecret`.
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+}
+
+impl fmt::Debug for ClientSecret {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(self.as_str())
+    }
+}
+
+impl Clone for Box<ClientSecret> {
+    fn clone(&self) -> Self {
+        (**self).to_owned()
+    }
+}
+
+impl ToOwned for ClientSecret {
+    type Owned = Box<ClientSecret>;
+
+    fn to_owned(&self) -> Self::Owned {
+        Self::from_owned(self.0.to_owned().into_boxed_str())
```

Comment:
Just noticed when re-reading that this does an unnecessary allocation & copy. Can be adjusted like the others: https://github.com/ruma/ruma/commit/e308c70aa75e8870ead8edc550174f23e6d705f6